cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/task/ProvisionTask.java

```diff
@@ -0,0 +1,202 @@
+                                                    cluster.getStatus()));
+    });
+    subtasks.put(ProvisioningOp.Status.POLLING_CREATE, subtask);
+
+    // RequestingDelete
+    subtask = new ClusterDeleteSubtask(provisioner, provisionerContext,
+                                       cluster -> Optional.of(ProvisioningOp.Status.POLLING_DELETE));
+    subtasks.put(ProvisioningOp.Status.REQUESTING_DELETE, subtask);
+
+    // PollingDelete
+    long taskStartTime = System.currentTimeMillis();
+    subtask = new ClusterPollSubtask(provisioner, provisionerContext, ClusterStatus.DELETING, cluster -> {
+      switch (cluster.getStatus()) {
+        case CREATING:
+          // this would be really weird, but provisioners can do whatever they want
+          return Optional.of(ProvisioningOp.Status.POLLING_CREATE);
+        case RUNNING:
+          // this would be really weird, but provisioners can do whatever they want
+          return Optional.of(ProvisioningOp.Status.INITIALIZING);
+        case NOT_EXISTS:
+          // delete succeeded, try to re-create cluster unless we've timed out
+          if (TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis() - taskStartTime) > retryTimeLimitSecs) {
+            // over the time out. Give up and transition to deprovisioned
+            provisionerNotifier.deprovisioned(programRunId);
+            return Optional.of(ProvisioningOp.Status.FAILED);
+          } else {
+            return Optional.of(ProvisioningOp.Status.REQUESTING_CREATE);
+          }
+        case FAILED:
+          // create failed, issue a request to delete the cluster
+          return Optional.of(ProvisioningOp.Status.REQUESTING_DELETE);
+        case DELETING:
+        case ORPHANED:
+          // delete failed or something went wrong, try to deprovision
+          provisionerNotifier.deprovisioning(programRunId);
+          return Optional.of(ProvisioningOp.Status.FAILED);
+      }
+      // should never get here
+      throw new IllegalStateException(String.format("Unexpected cluster state %s while polling for cluster state.",
+                                                    cluster.getStatus()));
+    });
+    subtasks.put(ProvisioningOp.Status.POLLING_DELETE, subtask);
+
+    // state for Initializing
+    subtask = new ClusterInitializeSubtask(provisioner, provisionerContext, cluster -> {
+      provisionerNotifier.provisioned(programRunId, initialTaskInfo.getProgramOptions(),
+                                      initialTaskInfo.getProgramDescriptor(), initialTaskInfo.getUser(),
+                                      cluster, initialTaskInfo.getSecureKeyInfo());
+      return Optional.of(ProvisioningOp.Status.CREATED);
+    });
+    subtasks.put(ProvisioningOp.Status.INITIALIZING, subtask);
+
+    // states for Failed and Created
+    subtasks.put(ProvisioningOp.Status.FAILED, EndSubtask.INSTANCE);
+    subtasks.put(ProvisioningOp.Status.CREATED, EndSubtask.INSTANCE);
+
+    return subtasks;
```

Comment:
This method is really long, better break it down into smaller chunk to make the intention clearer. E.g. the creation of the polling tasks could be separated into private methods.